## 4. 電卓代わりの計算ツール
日常のオフィス業務においてJupyter notebookを常時使用している状態を念頭に置いています．そのためには簡単に使えて便利であることを体感することが大事です．そういう観点で，最初に取り組む使い方は「電卓」です．<br>
Pythonを使うというとプログラムを書くことと思ってしまいますが，プログラミングに行く前にもっと単純だが役に立つ使い方があります．
プログラミングの経験の無い方は，プログラミング言語であるPythonの活用というと，どうしてもプログラミングの壁にぶち当たってしまいます．
その障壁を乗り越えるためにも，先ずはプログラミングを前提としない使い方からアプローチしていきます．

一般的な電卓の機能としては，四則演算と平方根ぐらいです．関数電卓だと数学的関数計算もできますが，まずは単純な電卓の代わりとしてJupyter notebookを使います．

### 四則演算
四則演算は，足し算・引き算・掛け算・割り算ですが，Pythonではどのように記述するか確認しましょう．


| 演算   | 数学的記法  | Python表記法 | 　　　計算例　　　  |
| :---:  | :---:       | :---:        | :-----------------  |
| 足し算 | $a+b$       | $a+b$        | $2+3=5$             |
| 引き算 | $a-b$       | $a-b$        | $2-3=-1$            |
| 掛け算 | $a\times b$ | $a*b$        | $2\times 3=6$       |
| 割り算 | $a\div b$   | $a/b$        | $2\div 3=0.666...$  |

それでは四則演算を順番にCodeセルで実施してみましょう．

*****
##### 足し算
足し算の練習です．次の通りCodeセルに入力して実行ボタンをクリックしてください．
```Python
1+2
```

In [1]:
1+2

3

入力セルの下に左端のOut[ ]とともに計算結果が表示されます．この部分を出力セルといいます．<br>
式を書くときに見やすくするために式の中ににスペースを挟んでも大丈夫です．
```Python
1 + 2
```

In [2]:
1 + 2

3

数字を記入するときに，整数か小数点付き数か気にする必要はありません．計算式の中に小数点付き数があると，結果も小数点付き数となります．
```Python
1 + 2.0
```

In [3]:
1 + 2.0

3.0

桁数の大きな数を扱う場合に３桁ごとにカンマで区切った数字を記載すると，Pythonでは配列として扱われてしまいます．<br>
４桁の数字の足し算の例として$1,000+2,000$を行います．この計算式をカンマ有りで入力してみます．
```Python
1,000 + 2,000
```

In [4]:
1,000 + 2,000

(1, 2, 0)

計算結果は(1,2,0)という配列になってしまいました．この式をPythonは次のように解釈しています．<br>
```Python
1,000 + 2,000 = (1, 000+2, 000)) = (1, (0+2), 0) = (1, 2, 0)
```
すなわち，3つの数値がカンマで区切られて並べられているものと解釈されます．ここで，000は0として扱われています．　<br>
0から始まる数字でも計算してくれそうに思われるので，同じようなケースとして10+00を実行してみます．
```Python
10 + 00
```

In [5]:
10 + 00

10

0から始まる文字を数値として扱ってくれそうなので，確認のためにもう一つ計算を行ってみます．今度は
```Python
10 + 02
```
を実行します．すると，次のようなエラーが発生します．

```
  File "<ipython-input-6-b189316209ef>", line 1
    10 + 02
          ^
SyntaxError: invalid token
```

今度は02が数値として認識されずエラーとなってしまいます．数字を書くときに全体の見栄えを揃えるために頭に0を付けて文字の長さを調整することがありますが，このようなトラブルがあることも理解しましょう．

桁数が大きな数値の表記法として指数表記があります．これは$a\dot\times 10^b$という形式での表現となります．例として百万の表記を見てみましょう．
英数字では$1,000,000=1\times 10^6$ですが，Pythonでは下記のように記述します．
```Python
1e6
```
これは「1掛ける10の6乗」と読めます．ここでeは大文字Eでも構いません．<br>
先ほどカンマでトラブルを起こした式$1,000+2,000$を指数表記で実施してみます．
```Python
1e3 + 2e3
```

In [6]:
1e3 + 2e3

3000.0

小数点以下の小さな数を表現する場合は指数部分の数値にマイナスの数字を入れます．例えば$0.001=1\times 10^{-2}$なので
```Python
1e-2
```
と記載します．以下のCodeセルで確認しましょう．

In [7]:
1e-2

0.01

ここに示した数値の表記法の注意事項は，足し算だけでなくPython全体で通用することですので，しっかり理解してください．

*****
##### 引き算
次は引き算について試してみましょう．Codeセルに次のように入れて実行します．
```Python
5-3
```

In [8]:
5-3

2

期待通りの結果になりました．
次に負の数を引いてみましょう．
```Python
3--4
```

In [9]:
3--4

7

この計算は3から負の数である-4を引いたので結果は7になります．正しい結果となっていますが，式が見にくくなっています．
見易さを考慮するならば，次のどちらかの書き方が良いでしょう．
```Python
3 - -4
3-(-4)
```
最初の書き方は，演算子のマイナスの周りには空白を入れて，符号のマイナスと数値は離さないで記載する方法です．2番目の方法は，括弧を使って明確な指示を出しています．この方法だと手間が増えますが，曖昧性が除去できます．
```Python
3-(-4)
```

In [10]:
3-(-4)

7

ちなみに，
```Python
3---4
```
を計算すると結果は-1になります．数学の作法としては馴染まない表現ですが，マイナス記号が符号反転の一項演算子として定義されているのでそのような振舞いとなります．すなわち，$3---4=3--(-4)=3-(-(-4))=3-4=-1$と解釈されます．

*****
##### 掛け算
掛け算はプログラム言語が発明された時からアスタリスク記号「\*」を使ってきました．
数学では$a\times b$とするか$a \cdot b$と記載しますが，Pythonでは$a\ast b$とします．
次の計算を行って動作を確認します．
```Python
2*3
```

In [11]:
2*3

6

負の数の掛け算も行ってみましょう．見やすくするためにアスタリスクの前後に空白文字を挟みます．
```Python
2 * -3
```

In [12]:
2 * -3

-6

*****
##### 割り算
割り算の演算子についても他の言語と同じくスラッシュ「/」を用います．
例として$3\div 2$を計算してみます．次のように入力して実行キーをクリックします．
```Python
3/2
```

In [13]:
3/2

1.5

このように整数どうしの計算でも結果が有理数ならば小数点付き数で答えが返ります．
もう一つ別の計算してみます．
```Python
2 / -3
```

In [14]:
2 / -3

-0.6666666666666666

この場合，割り切れないので小数点以下6が無限に続くのですが，このように途中で打ち切られます．
これはデータの持つ精度に限界があり，特に割り算で計算誤差が発生する可能性があります．

割り算では，計算誤差の他にゼロで割るエラーにも気を付ける必要があります．ここでは割る数を明示的に与えているので0でない数にできますが，変数を使いだすと変数の状態によってはゼロで割る可能性が出てきます．
小さくてもゼロでない数を扱っていても，次のように精度の問題でゼロと見なされてしまいます．
```Python
1e-324
```

In [15]:
1e-324

0.0

このように，非常に小さい数になりそうなケースに気を付けましょう．

これでPythonで四則演算を行えるようになりました．Jupyter notebookを使用すると，これまで計算した結果が文章として残っています．電卓での計算は一過性なので，タイプミスがあってもやり過ごす危険があります．さらに電卓での検証では，同じ計算を何度も繰り返す作業になりますが，Jupyter notebookでっは文章の確認作業となり，作業効率に大きな差が出ます．

これからのオフィスワークでは常にJupyter notebookを立ち上げておいて，ちょっとした計算でもログを残すようにしましょう.

*****
### 冪乗（累乗）
同じ数を繰り返し掛ける演算を累乗と言います．累乗は$a^b$のように記述しますが，$b$を自然数から実数へ拡張した演算を冪乗と言います．
次の式は累乗の例となります．

> $a^5=a\times a\times a\times a\times a$

このように，まず$n$が自然数の場合の$a^n$が定義できます．
次に$a^n$となる$x$を$a$のn乗根として$a^{1/n}$が定義されます．
さらに$a^m$のn乗根を$(a^m)^{1/n}=a^{m/n}$と表すことから，正の有理数$p$に対して$a^p$を定義することができます．これを連続的に極限を取ることによって正の実数$r$に対して $a^r$が定義されます．

Pythonでは冪乗$a^b$を次のように記述します．
```Python
a**b
```
それでは，幾つかの例で確認しましょう．

> $2^2=4$<br>
> $2^{-2}=1/4$<br>
> $2^{1/2}=\sqrt{2}$<br>
> $(-2)^{1/2}=\sqrt{2}\cdot i$

それぞれ，次にように入力します．
```Python
2**2
2**(-2)
2**(1/2)
(-2)**(1/2)
```

In [16]:
2**2

4

In [17]:
2**(-2)

0.25

In [18]:
2**(1/2)

1.4142135623730951

In [19]:
(-2)**(1/2)

(8.659560562354934e-17+1.4142135623730951j)

３番目の例は平方根の計算になっています．すなわち$\sqrt{a}$はa\*\*(1/2)で計算できます．ちなみに平方根を求める関数も存在します．関数で記述する方が記入間違いが起こりにくいので，実用上は関数で計算します．

４番目の例ですが，正確には$\sqrt{-2}=\sqrt{2}\times i$となりますが，この計算結果は$8.66\times 10^{-17}+1.41\times i$となっています．この実数部はゼロのはずですが計算誤差が出ています．また虚数部の表現ですが，虚数単位$i$が$j$で表されます．

さて，数値には整数のint型と小数点付き数のfloat型がありましたが，複素数はcomplex型となります．type()関数で確認しておきましょう．
```Python
type(1+1j)
```

In [20]:
type(1+1j)

complex

*****
### 複雑な計算式
ここまでは単純な四則計算を行いましたが，プログラム言語なので複雑な計算式を1行で計算することもできます．<br>
例えば，3名で複数の料理をシェアしたので，それを等分に分けることにします．

| 品目 | 単価 | 個数 |
| :---: | :---: | :---: |
| 枝豆   | 250 | 4 |
| サラダ | 340 | 1 |
| 唐揚げ | 420 | 2 |
| 餃子   | 440 | 3 |

合計金額に8%の税金がかかるものとします．合計金額を3等分します．
```Python
(250*4 + 340*1 + 420*2 + 440*3)*1.08/3
```

In [21]:
(250*4 + 340*1 + 420*2 + 440*3)*1.08/3

1260.0000000000002

と言うことで一人当たり1260円の支払いとなります．
現実的にはお店側から合計金額の3780円が告げられるので，このような計算をする必要はありませんが，分かりやすい例として挙げました．

##### 演算の優先順位
複雑な計算式を入力する場合は，計算の優先順位を知っている必要があります．ここで計算の優先順位について再確認しておきましょう．

| 優先順位 | 演算種類 | 　　　例　　　 | 
| :---: | :--- | :---: |
| 1 | 括弧内 ( ) | |
| 2 | 冪乗，冪根 | $a^b,a^{1/b}$ |
| 3 | 掛け算，割り算 | $a\times b, a\div b$ |
| 4 | 足し算，引き算 | $a+b, a-b$ |

簡単な例として

> $3+2\times 5=13$<br>
> $(3+2)\times 5=25$

について確認します．
```Python
3+2*5
(3+2)*5
```

In [22]:
3+2*5

13

In [23]:
(3+2)*5

25

別な例を見てみましょう．今100万円の資金があり年利2％の10年物の社債を購入したとすると，10年後は幾らになるでしょうか．<br>
複利計算なので計算式は，

> $元金\times (1+年利)^{年数}$

となります．
```Python
1000000 * (1+0.02)**(10)
```

In [24]:
1000000 * (1+0.02)**(10)

1218994.4199947573

10年後には約122万円弱になります．
このように計算式を巧みに利用すれば，電卓を超える便利なツールとなります．
そのとき問われるスキルは，計算式を作り出す能力です．目の前の課題に対してどのように定式化するかは見逃されやすい部分ですが，定式化された問題を解くことよりも重要です．

*****
### 商と余り
通常の割り算では，どんなには数が出ても割り切ってしまいます．しかし，結果が整数で欲しい場合もあります．そのような場合に商と余りを求める演算が用意されています．

> $10 \div 3 = 3 余り 1$

商と余りを求めるには次の演算子を使います．

| 演算 | 記法 | 　　例　　 |
| :---: | :---: | :---: |
| 商   | $a//b$ | $10//3=3$ |
| 余り | $a\%b$ | $10\%3=1$ |

次の式を実行しましょう．
```Python
10//3
10%3
```

In [25]:
10//3

3

In [26]:
10%3

1

商と余りの計算が必要になる頻度は少ないですが，このような計算もできることを頭の片隅においてください．<br>
例としては，215分は何時間何分か調べる場合，
```
215 // 60 = 3
215 % 60 = 35
```
によって求まります．ただし，Pythonにおいて時間計算は専用のライブラリーを使うことが一般的です．

*****
### 変数の活用
ここまで，Jupyter notebookでPythonによって計算をする方法について見てきましたが，これだけだとプログラミング言語の利点を使っていません．
そこで変数の使い方についてマスターしましょう．
これまで，Codeセルに
```Python
1+2+3+4+5
```
というように入力していました．

In [27]:
1+2+3+4+5

15

このように出力セルに計算結果が表示されます．しかし，この結果を利用して別な計算をしたい場合は値を再入力しなければなりません．
値を何度も利用するために，計算結果を変数に格納します．<br>
ここで変数として「total」という名称を用います．この変数totalに上述の計算結果を代入します．
変数への値の代入は「=」記号を用いて行います．
```Python
total = 1+2+3+4+5
```


In [28]:
total = 1+2+3+4+5

代入文を実施した場合，出力セルには何も表示されません．代入された値を確認するにはCodeセルに変数名を記入して実行キーをクリックします．
```Python
total
```

In [29]:
total

15

変数名totalは，この代入文で初めて出てきたのでこの代入文によって定義されています．
Pythonによるプログラミングにおいて変数を使用する前に宣言する必要はありません．

変数の値を使う場合は，値の代わりに変数名を記載すれば実行時に値を代入して計算してくれます．
```Python
total / 5
```

In [30]:
total / 5

3.0

まず，ここまでの内容でPythonを電卓代わりに使うことが可能となりました．
さらに，過去の計算結果が実行ログとして残っているので，ミスの防止と検証が飛躍的に向上します．

##### 変数の命名センス
変数名についてのPythonのプログラミング上のルールは，
- 英数字とアンダースコアが使える．
- 数字は先頭文字に使えない．

ですが，慣例的な命名作法もあります．例えば，アンダースコアで始まる変数名はオブジェクトの内部だけで使用する，などです．
しかし，それらの作法について今のところ必要ないので，本格的なプログラムを作成する段階で確認してください．

ここでは，変数名称を付けるときの最低限のセンスについてガイドします．

- 変数名は英語で意味の通じる名前にする．
- 既存の関数名などを上書きしないように気を付ける．
- 英単語2つ以上を組み合わせた変数名を作る．
- 英単語の接続は単語の先頭を大文字にする方法か，あるいは，アンダースコアで繋ぐ方法があり，なるべくどちらかに統一する．

変数を使うということは，そのデータに何らかの意味があるはずです．その意味に沿った名前を英語で付けましょう．
たまに，変数名に日本語のローマ字表記を用いる人がいますが，グローバル社会に適合するには，変数名やコメントは全て英語にすべきです．コメント文では文法を気にする必要はありません．思い浮かべた言葉の英語が分からなければ翻訳ツールなどを使って検索してください．
ただし，文字数の長い単語について省略形を使うことは問題ありません．例えはdepertmentをdeptと記載しても意味は分かります．

2つ以上の英単語を繋げる2通りの例を見てみましょう．今、部屋のサイズをある変数に入れたいと考えます．「部屋のサイズ」なので英語では「room size」ですが，これを頭大文字繋ぎとアンダースコア繋ぎで書くと次の通りになります．

- roomSize
- room_size

両方ともよく使われます．どちらにするかは好みの問題ですが，どちらかに統一することをお勧めします．

*****

ここまで，Pythonで簡単な計算を実施する方法を学びました．